### ISSUE TITLE
OSM Map loading slower than it used to be

### ISSUE THREAD

**Issue by Lineflyer** [cgeo/cgeo#9292]
Taken from https://github.com/cgeo/cgeo/issues/9270#issuecomment-716890164 reported by @MagpieFourtyTwo as a possible regression after fixing that issue:

Map loading is way slower than ... a few days ago ;)
e. g.: with Show as list it takes up to 5 secs until the map is loaded completely (even with just BaWü tile selected),
which used to happen during a wink of the eye before

**Comment by MagpieFourtyTwo** [comment]
@eddiemuc 
> is the map loading slower than before only when you select a single map file or also when you use "combined map"?

``Combined`` is not significantly faster. While it depends on how many map files are involved: If all caches to be displayed are on one, it *may* be a little bit faster, but if there are more than one map files involved, it's even slower. Although this may also be a matter of how big the displayed area is. Have to check more detailed, once I have more time.

**Comment by SammysHP** [comment]
If map loading is the bottleneck, we should use the file that is already opened for rendering and in case of combined maps only when the popup is opened.

**Comment by eddiemuc** [comment]
> If map loading is the bottleneck, we should use the file that is already opened for rendering and in case of combined maps only when the popup is opened.

Usage of same mapfile for map attribution and rendering was the cause for the map rendering problems solved last night. Can't really tell you why, but somehow underlying file got closed as it was time to render the map (no, I did not call close on MapFile after retrieving attribute info :-)). Might also be a problem with how we reuse MapSource instances in our code.

Currently for combined map the mapfiles are opened asynchronously to get attribution info out. For single maps it is currently synchronous. I can change this to asynchronous as well, but if this is the cause of the problem then opening of combined maps should currently work as fast as before, and according to @MagpieFourtyTwo it doesn't.

**Comment by eddiemuc** [comment]
Created PR #9299. This PR makes calc to mal attribution asynchronous for single offline maps. In hope that this is the problems cause

**Comment by MagpieFourtyTwo** [comment]
To consolidate my background I went through #6150 again, trying to find some entry points for further testing ...

Reading this: "to actually get the attribution information out of them, I have to open them" I thought it might possibly be a factor when having plenty of files, so I reduced my set from 17 to 3 map files - but to no avail. With regard to speed this reduction makes no difference at all. Even with just 1 map file it's as slow as with 17. This info just in case ...

Then I moved the remaining map file from external SD to internal memory, but this did not make a difference neither. As expected, to be honest, cause the Huawei NM cards are amazingly fast, sometimes even faster than internal memory.

Last but not least I stumbled upon the popup that appears "On click:" ... over here it reads "Combined (17 maps, 2 attributions): click for details" (cause apart from OpenAndroMaps I have a world map from a different source installed) - is this just a kind of "dirty workaround" to solve the "not enough room for all contributions" problem or would such a message be sufficient? I. e.: Would it be ok to show (and create) the attribution text after a click only? This would at least save the time on every single map opening to get the data for the attribution(s) ... or, would it perhaps be a possibility to cache the last attribution (in DB?), to just check it just now and then? Cause I guess, the map setup will not change *that* often, to make it necessary to recreate it from scratch on every single map opening ...? And if one wants to be on the safe side, the modification date of the map folder can perhaps be stored together with the cached attribution ...?

**Comment by MagpieFourtyTwo** [comment]
Ah forgot to mention another observation: Even when I just switched to Android's homescreen and back to c:geo via "Recent Apps", the map gets cleared and redrawn completely, although it was already there ... was this always the case? Cause if so, map display was *way* faster before, cause I did not notice such a redraw since - and right now it takes the mentioned 4 to 5 secs, during which one sixth of the screen gets "painted" after the other ...

BTW: If I move the map in a way that the upper and lower "row" have just half of the height of the others, the order in which the map is painted is like this ...

 7 | 8 
---|---
**3** | **4**
 **1** | **2**
 **5** | **6**

... where - and this is what I think may probably be remarkable - tiles 7 and 8 always take the longest time until they appear: While tiles 1 to 6 appear with just a "slight" delay of perhaps a quarter of a second (which is of course everything else than "slight" ;) ), it takes about a whole second until 7 and (with another quarter of a second) eventually 8 appears.

**Comment by eddiemuc** [comment]
Hi @MagpieFourtyTwo this map attribution thing is my first code change ever with regards to map topics. Sorry for the problems potentially caused by my non-knowledge of the (non-obvious) interactions between different code parts...

> Reading this: "to actually get the attribution information out of them, I have to open them" I thought it might possibly be a factor when having plenty of files, so I reduced my set from 17 to 3 map files - but to no avail. With regard to speed this reduction makes no difference at all. Even with just 1 map file it's as slow as with 17. This info just in case ...

I would expect that number of files does not make any difference since in case of "combined" the map attribution is read asynchronously to the GUI flow. I still don't get how the change slows down map reading in general...

> Last but not least I stumbled upon the popup that appears "On click:" ... over here it reads "Combined (17 maps, 2 attributions): click for details" (cause apart from OpenAndroMaps I have a world map from a different source installed) - is this just a kind of "dirty workaround" to solve the "not enough room for all contributions" problem or would such a message be sufficient? 

Yes, this is a way to deal with room restrictions. But why do you find it dirty?

> I. e.: Would it be ok to show (and create) the attribution text after a click only? 

As I said, the creation is already asynchronous. It must be done upfront though because otherwise I can't find out whether all maps have the SAME map attribution (in this case, there is no "click" message but attribution is directly displayed)

> would it perhaps be a possibility to cache the last attribution (in DB?), to just check it just now and then? Cause I guess, the map setup will not change _that_ often, to make it necessary to recreate it from scratch on every single map opening ...? And if one wants to be on the safe side, the modification date of the map folder can perhaps be stored together with the cached attribution ...?

Well everything is possible. But this would be a very complicated solution, I still hope there is an easier way

**Comment by eddiemuc** [comment]
> Ah forgot to mention another observation: Even when I just switched to Android's homescreen and back to c:geo via "Recent Apps", the map gets cleared and redrawn completely, although it was already there ... was this always the case? Cause if so, map display was _way_ faster before, cause I did not notice such a redraw since - and right now it takes the mentioned 4 to 5 secs, during which one sixth of the screen gets "painted" after the other ...

I don't think we have a direct influence on how the MapView (which is coming from mapsforge) paints the map, in which order it paints the tiles or such things

**Comment by eddiemuc** [comment]
@MagpieFourtyTwo one thought crossed my mind. Is the painting of the map slower/faster for you if you open a piece of map where there are no caches? Or where there are caches but none which are stored in your local db?

**Comment by MagpieFourtyTwo** [comment]
Due to #9293 it's currently a little bit hard to get a completely "fresh" map window spanning more than one "tile" (i. e. one sixth of a screen, while map is painted), so I can just check map refresh while moving the map around or by pressing Android's home button, in order to return to c:geo via "Recent Apps", which currently (unnecessarily) redraws the map completely (which I in fact would call a bug, but in our case this bug is a feature ;) ).

So I created this scenario:
1. open Live Map
2. disable Live Update
3. scroll to a region where no caches are displayed
4. press Android's home button
5. return to c:geo via "Recent Apps"

And in this scenario, it's as slow as with a bunch of caches on display.
To visualize the descriptions in [my previous comment](https://github.com/cgeo/cgeo/issues/9292#issuecomment-717584130) (and to proof that it's really 4 to 5 secs ;) ) I added a screen recording [2020_10_28_103020.mp4.zip](https://github.com/cgeo/cgeo/files/5452839/2020_10_28_103020.mp4.zip), which may give you an impression (had to zip it, as github does not accept mp4 files) ...

Looking at this file it seems to me that first all the entirely shown tiles are drawn and then the partial ones, following the strategy right/left / bottom center/right/left / top center/right/left ... interesting.  ;)  Never got aware of this before, which may even be an additional indicator for how fast it was before ...

**Comment by moving-bits** [comment]
I'm not convinced that the cause for this delay is missing asynchronicity. I've added a bit of measuring times on current master, and retrieving the map attribution takes about 2ms in the emulator, and is done only once per file. So the technical process of retrieving the attribution data does not seem to be the cause, there has to be a different thing causing the delays.

**Comment by Lineflyer** [comment]
@MagpieFourtyTwo 
After looking on your video...what device are you using. 
However: For me on such low zoom it has always been like that. I noticed however differences when it comes to which map I use. Mapsforge tends to be slower for me than Openandromaps (maybe due to the file size itself?).
Additionally it even more depends on the theme you are using up to my experience.

Coming back to the topic:
I do not see a difference in loading speed of OSM offline maps during the last weeks.

**Comment by MagpieFourtyTwo** [comment]
Any chance to get a PR apk with additional debug / timing info to find the time consuming stuff?

**Comment by MagpieFourtyTwo** [comment]
> However: For me on such low zoom it has always been like that.

Over here it was never ever that slow. Not even with the "old" Xperia Z5 Premium,
which was not even nearly as performant as my current device is: Huawei P30 Pro.
Right now even GSat is faster - and this always was the slowest.

**Comment by MagpieFourtyTwo** [comment]
Additionally disabled all "Theme Options" - didn't change anything, neither.
So, I guess, Theme can be excluded as reason ...
Apart from that I did not change anything with regard to the map.

**Comment by MagpieFourtyTwo** [comment]
> on such low zoom

Now that #9293 is fixed, I could more easily compare different zoom levels again.
And - yes, it (of course) makes a difference, if the zoom is "2 km" or "50 m", but not as much as one would expect:
With "2 km" it needs the described 4-5 secs, but even with "50 m" it's about 3 secs.
While it takes the most time until the first tile starts to be drawn. Then the following 5 come faster than with "2 km".
But still: Have never ever seen a complete blank map view, like right now, immediately after opening the map view.

**Comment by eddiemuc** [comment]
Hi @MagpieFourtyTwo ,

I tried on my private phone (last release 2020.10.10 installed) how long map update needs when I switch themes between Elevate.xml and Elements.xml (I use OpenAndroMaps). Result is: while map update with Elevate.xml is "lightning-fast", with "Elements.xml" it is as slow as shown in your video. So I assume that this is where the problem lies.

Also from your video, I can see that map attribution is not the problem - because at the time the attribution line is shown as we see there (with the "click for details" message) the whole attribution has already been parsed and stored.

Regarding additional debug info: I am afraid this is not possible because the whole map rendering stuff is done by mapsforge library, which to us is a 3rd party lib where we can't enter additional log messages and such.

Currently I fear there is nothing we can do to speed things up apart from selecting less detailled map themes

![image](https://user-images.githubusercontent.com/6909759/97537636-680c0c00-19bf-11eb-85b8-f020cb63c01c.png)

**Comment by MagpieFourtyTwo** [comment]
As written before change of map theme does not significantly speed up things.  No matter if I use "Elements", "Elevate" or even "(default)", it always takes nearly the same time, and even disabling all "Theme Options" changes nothing, at least not in a significant way.  Really strange ...

As for the attribution: You're right, the attribution is there in absolutely no time. So this seems to be out ... assumed that the attribution does not close the files currently (or previously) in use ...

However, there are way to many redraws ... esp. when re-entering the map there is in fact absolutely no need to redraw it ... can this possibly be a consequence of the removed context snapshot on leaving the map?

BTW: How many map files have you installed and which map versions are you using? I updated mine last time on 05.10.2020.

**Comment by eddiemuc** [comment]
> BTW: How many map files have you installed and which map versions are you using? I updated mine last time on 05.10.2020.

I use version 4 maps and have currently three map files on my device:
- germany as of June 2020
- netherlands and belgium as of August 2020

(You can probably guess now where we went on vacation this summer holiday :-))

**Comment by eddiemuc** [comment]
Additional note: with regards to map display speed, I don't notice an difference between choosing "combined map" or a single map file

**Comment by MagpieFourtyTwo** [comment]
Ok, I'm using OpenAndroMaps, too - so, could you perhaps please update your map files and check again afterwards?
Perhaps it's just the new(er) map file version ...?

> Additional note: with regards to map display speed, I don't notice an difference between choosing "combined map" or a single map file

Me neither.

**Comment by MagpieFourtyTwo** [comment]
BTW: Are there (or do you notice) any problems with download.cgeo.org right now?
Although in PC's browser the download of cgeo-nightly.apk works perfectly fine,
with Firefox for Android it takes ages ... if it downloads at all ...
Moreover, when starting c:geo, the version check says "New nightly build available" although I'm already on the latest one ...
Just don't get what's going on there ...

**Comment by Lineflyer** [comment]
> Moreover, when starting c:geo, the version check says "New nightly build available" although I'm already on the latest one ...
> Just don't get what's going on there ...

I just checked the status notifier and the nighty builds on CI. I did not notice any issue.
Maybe you are still one nightly behind as I triggered a manual NB1 build yesterday late evening?

**Comment by eddiemuc** [comment]
> Ok, I'm using OpenAndroMaps, too - so, could you perhaps please update your map files and check again afterwards?
> Perhaps it's just the new(er) map file version ...?

I noticed just now that I used another zoom level than you did in your mp4. When I use the same zoom level, then my map update speed is about the same as the one in your clip, also when I use "elevate.xml". I can make it a bit slower/faster by selecting more or less of the Theme options, but overall I think I have the same experience as you.

**Comment by fm-sys** [comment]
> with Firefox for Android it takes ages ... if it downloads at all ...

I made the same experience while downloading your mp4 from above with the recent Firefox for Android. So it is probably more a Firefox bug...

**Comment by MagpieFourtyTwo** [comment]
Yep, was aware of the extra Nightly and in fact just wanted to get rid of the message box,
in order to see when the next *really* new one arrives.
Finally "removed" the box by downloading via PC, but download via Firefox for Android still does not work reliable.

@fm-sys: Yep, very likely. Most regretfully I do not have another browser ... perhaps I might have to take a look at Kiwi after all ...

**Comment by geocachermgo** [comment]
> > with Firefox for Android it takes ages ... if it downloads at all ...
> 
> I made the same experience while downloading your mp4 from above with the recent Firefox for Android. So it is probably more a Firefox bug...

Same for me, I had to use Mi browser to download. My guess was that it's an intentional security feature.

**Comment by MagpieFourtyTwo** [comment]
Installed Kiwi browser.
And to be honest: Although I'm an absolute Firefox addict for decades ... the more I use it, the more I like it. Even got Tampermonkey and GClh installed and running again, which is currently impossible with FF4Android.

**Comment by MagpieFourtyTwo** [comment]
Anyway, back2topic: If the draw speed cannot be tweaked (due to closed Mapsforge lib), is there any chance to eliminate the unnecessary redraws? This would at least provide some relief ...

**Comment by moving-bits** [comment]
Won't fix. Old map implementations are deprecated as of release 2025.07.17 and superseded by UnifiedMap.
